File: problem3.py
def primes():
    yield 2
    ps = [2]
    n = 3
    while True:
        for p in ps:
            if n % p == 0:
                break
            if p**2 >= n:
                ps.append(n)
                yield n
                break
        n += 2

def largestPrime(n):
    for p in primes():
        while n % p == 0:
            n //= p
        if n == 1:
            return p

print(largestPrime(600851475143))

File: problem3.hs
primes = 2:filter f [3,5..]
    where f n = all ((/= 0) . (mod n)) (takeWhile ((<= n) . (^ 2)) primes)

largestPrime n = checkNext n primes
    where
        checkNext n (p:ps) = if n == p
            then n
            else if n `mod` p == 0
                then checkNext (n `div` p) (p:ps)
                else checkNext n ps

main = putStrLn $ show $ largestPrime 600851475143
